Visibility Simulator Overhaul

steven-murray opened the pull request:
This PR should do the job of the previous #162 and #163, and hopefully not change EVERY file.

Fixes #154

* It fixes `VisCPU` and `HealVis` to use the new setup
* Adds `UVSim` simulator.

@hughbg can you check especially if the call to correct_point_source_pos is good? And also if polarization is being used correctly?

r-pascua commented on hera_sim/visibilities/vis_cpu.py:
```diff
@@ -315,38 +227,48 @@ def _base_simulate(self, crd_eq, I_sky):
             beam_list = [
-                self.beams[np.where(self.beam_ids == ant)[0][0]]
-                for ant in self.ant_list
+                data_model.beams[np.where(data_model.beam_ids == ant)[0][0]]
+                for ant in data_model.ant_list
```
This isn't going to work with the way the `beam_ids` array is constructed, since `beam_ids` won't necessarily have a unique entry for each antenna.

I think I'm a little too tired (and my head too stuck in the mist of edge cases) to figure out the right way to do this right now, but the thing I'm worrying about is the case where the antenna numbers don't monotonically increase in the `UVData` object. I'm mainly concerned about ensuring that we get the ordering right on all the moving parts here, so that the beams get matched with the right antennas and the visibilities returned by the simulator match the blt-ordering imposed on the data array.

steven-murray replied:
Yup. I'm with you. I'm wary of all of this, because I know there's a few moving parts. Perhaps we should chat live about it, and maybe rope in @mkolopanis  / @bhazelton .

r-pascua replied:
I think a live chat would be good. Maybe after next week's Validation meeting?

steven-murray replied:
Sounds good.

r-pascua reviewed (commented):
This is looking really great! A huge improvement from where it was before. I made a handful of comments, some that I think are important to get to, others not so much.

I'm running out of mental steam right now, but I definitely think it would be worthwhile for us to sit down and brainstorm the right way to handle how various arrays are sorted, because getting that wrong leads to nasty silent bugs.

Anyway, thank you for your work! I really like the direction this is going.

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -13,374 +12,129 @@
-        point_source_flux=None,
-        nside=2 ** 5,
-        validate=True,
+        *,
+        uvdata: UVData,
+        sky_model: SkyModel,
```
The idea is to make sure they're always passed as keywords. This is good for future-proofing the API (in case we change parameter ordering in any way).

r-pascua replied:
Ah, gotcha! I think this is a very good move, then.

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -13,374 +12,129 @@
+        *,
+        uvdata: UVData,
+        sky_model: SkyModel,
+        beam_ids: Dict[str, int] | None = None,
+        beams: BeamListType | None = None,
```
Apparently this is the preferred syntax in the annotations you import from `__future__`.

r-pascua replied:
Huh, ok

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -13,374 +12,129 @@
+            self.uvdata.Nants_data
+        )  # NOT Nants because we only want ants with data
+
+        assert isinstance(self.uvdata, UVData)
```
Fair enough. Done in aca8de5

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -13,374 +12,129 @@
-            lon=point_source_pos[:, 0] * units.rad,
-            lat=point_source_pos[:, 1] * units.rad,
-            nside=nside,
+        else:
+            self.beam_ids = np.array(beam_ids, dtype=int)
```
I was actually considering whether we instead preferred to keep the beam_ids dict from pyuvsim instead of the list that we turn it into. Or perhaps keep both around on the instance. We can allow both as input at least, good point.

The `ModelData` class is going to be used by the user to build up a `VisibilitySimulation`.

steven-murray replied:
I've updated accept the dict input. Will await your ideas on whether we should keep the dict format around (or switch to it).

steven-murray replied:
Another thing to think about: in the the `from_config` method, we do this to get the `beam_ids` from the dict:

```python
nms = list(uvdata.antenna_names)
tmp_ids = np.zeros(len(beam_ids), dtype=int)
for name, beam_id in beam_ids.items():
    tmp_ids[nms.index(name)] = beam_id
beam_ids=tmp_ids
```

Instead of simply `np.array(list(beam_ids.values()), dtype=int)`. I wonder if we did this for a reason -- perhaps we were ensuring ordering for pre-py36, where dicts weren't ensured to have a particular ordering?

steven-murray replied:
Just to be sure, I changed the conversion in `_process_beams` to take account of this ordering, but simplified the above into this:

```python
beam_ids = np.array(
    [beam_ids[nm] for nm in self.uvdata.antenna_names], dtype=int
)
```

r-pascua replied:
Nice, this is much more compact. I'm not sure why it was done the way it was before, but this looks like it achieves the same thing.

As for whether we should use the dict format from `pyuvsim` or not... I think I need to have a better picture of how everything works before weighing in on that.

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -13,374 +12,129 @@
+        assert isinstance(self.beam_ids, np.ndarray)
+        assert self.beam_ids.dtype == int
+        assert self.beam_ids.max() < self.n_beams
+        assert len(self.beam_ids) == self.n_ant
```
I agree. I've now done this.

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -13,374 +12,129 @@
+
+        self.sky_model = sky_model
+        self.sky_model.at_frequencies(self.freqs * units.Hz)
+        assert isinstance(self.sky_model, SkyModel)
```
Good catch. Fixed.

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -415,3 +169,112 @@ def write_config_file(
+
+    @cached_property
+    def ant_list(self) -> np.ndarray:
+        """An orderd list of active antenna numbers."""
```
Done, thanks.

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -415,3 +169,112 @@ def write_config_file(
+class VisibilitySimulator(metaclass=ABCMeta):
+    """Base class for all hera_sim compatible visibility simulators."""
+
+    #: Whether this particular simulator has the ability to simulate point
+    #: sources directly.
+    point_source_ability = True
+
+    #: Whether this particular simulator has the ability to simulate diffuse
+    #: maps directly.
+    diffuse_ability = False
```
Yeah, that's true. Then they just have to set this attribute properly. I think most simulators assume input point sources by default.

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -415,3 +169,112 @@ def write_config_file(
+        """Simulate the visibilities."""
+        pass
+
+    def validate(self, data_model: ModelData):
+        """Check that the data model complies with the assumptions of the simulator."""
+        pass
```
Yeah the default `validate()` is to do nothing. Many simulators might not have to do any validation (i.e. they might allow any kind of data as input).

r-pascua replied:
Cool, thanks for the clarification!

steven-murray commented on hera_sim/visibilities/healvis_wrapper.py:
```diff
@@ -36,45 +37,45 @@ class HealVis(VisibilitySimulator):
+        In addition to standard parameter restrictions, HealVis requires a single beam
+        for all antennae.
+        """
+        assert model_data.n_beams == 1
```
Done.

steven-murray commented on hera_sim/visibilities/healvis_wrapper.py:
```diff
@@ -36,45 +37,45 @@ class HealVis(VisibilitySimulator):
             gauss_width = None
             if old_args["type"] == "gaussian":
                 if old_args["sigma"] is None:
                     raise NotImplementedError(
-                        "Healvis does not permit " "gaussian beam with diameter."
+                        "Healvis does not permit gaussian beam with diameter."
                     )
```
Yeah, I think this was something that happened in two stages. I've simplified it now.

steven-murray commented on hera_sim/visibilities/healvis_wrapper.py:
```diff
@@ -103,25 +94,30 @@ def sky_model(self):
-            2 * cnst.k_B.value
-        )
-
+        if sky_model.stokes.unit == units.Jy / units.sr:
+            intensity = 10 ** -26 * sky_model.stokes[0].T.value
+            intensity *= (
+                cnst.c.to("m/s").value / sky_model.freq_array.to("Hz").value
+            ) ** 2 / (2 * cnst.k_B.value)
+        elif sky_model.stokes.unit == units.K:
+            intensity = sky_model.stokes[0].T.value
+        else:
+            raise ValueError(
+                f"Units of {sky_model.stokes.unit} are not compatible with healvis"
+            )
```
I like that. Done.

steven-murray commented on hera_sim/visibilities/healvis_wrapper.py:
```diff
@@ -146,15 +142,12 @@ def _simulate(self):
+            obs.make_visibilities(sky, Nprocs=self._nprocs, beam_pol=pol)[0]
+            for pol in data_model.uvdata.get_pols()
+        ]
         visibility = np.moveaxis(visibility, 0, -1)
 
         return visibility[:, 0][:, np.newaxis, :, :]
```
OK, yeah I see it. This will have to be fixed before merging.

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -1,10 +1,9 @@
 """Module defining a high-level visibility simulator wrapper."""
-from __future__ import division
-import warnings
+from __future__ import division, annotations
```
The division shouldn't be there -- removed it now. The `annotations` are important for some nice features that are coming in py3.10+ (like using a class as a type within its own body, and using `|` for `Union`).

r-pascua replied:
Ah, interesting, ok! Thanks for clearing that up!

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -415,3 +169,112 @@ def write_config_file(
+                    np.zeros_like(hmap),
+                    np.zeros_like(hmap),
+                ]
+            )
+            * units.Jy
+            / units.sr,
```
Yeah. I tried to change it back but it wouldn't have it. 🤷🏻

r-pascua replied:
Well, if the robots say that's the way to do it...

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -415,3 +169,112 @@ def write_config_file(
+        """Write pertinent details of simulation to the UVData's history."""
+        class_name = self.simulator.__class__.__name__
+        self.uvdata.history += (
+            f"Visibility Simulation performed with hera_sim's {class_name} simulator\n"
```
Yes.

steven-murray commented on hera_sim/visibilities/simulators.py:
```diff
@@ -415,3 +169,112 @@ def write_config_file(
+
+    def simulate(self):
+        """Perform the visibility simulation."""
+        self._write_history()
```
Hmmmm good point. I'll change it to afterwards.

steven-murray commented on hera_sim/visibilities/vis_cpu.py:
```diff
@@ -48,18 +48,21 @@ class VisCPU(VisibilitySimulator):
     ):
 
-        assert precision in (1, 2)
+        assert precision in {1, 2}
```
Hmmm, that'll have to be coordinated with the `vis_cpu` package.

r-pascua replied:
Ah, ok.

steven-murray commented on hera_sim/visibilities/vis_cpu.py:
```diff
@@ -92,101 +95,40 @@ def __init__(
         # N(N-1)/2 unique cross-correlations + N autocorrelations.
-        if len(self.uvdata.get_antpairs()) != N * (N + 1) / 2:
+        if (
+            len(data_model.uvdata.get_antpairs())
```
Are you sure this is guaranteed to be the same? I don't have enough experience with pyuvdata, but I know there's sometimes differences between active/non-active antennas etc.

r-pascua replied:
I haven't extensively tested it, nor have I looked at the code for how this attribute is retrieved, so I can't say for sure. If we're being cautious, then I think we're better off leaving as-is. (All I've done is a shallow check where I select only a few antennas and confirm that the `Nbls` attribute matches the length of the returned antenna pairs.)

steven-murray replied:
OK, looking at pyuvdata source code, it looks like `Nbls` is always correct. I'll change the code.

steven-murray replied:
(Done)

steven-murray commented on hera_sim/visibilities/vis_cpu.py:
```diff
@@ -92,101 +95,40 @@ def __init__(
-        if len(self.uvdata.data_array) != len(self.uvdata.get_antpairs()) * len(
-            self.lsts
-        ):
+        if len(data_model.uvdata.data_array) != len(
+            data_model.uvdata.get_antpairs()
+        ) * len(data_model.lsts):
```
I think so. That's why they have a blt axis instead of splitting that into two axes.

r-pascua replied:
Fair enough. Do you know if we can get the same check by comparing, say, `Nblts` and `Nbls * Ntimes`? I'm not familiar enough with using `UVData` objects set up like that to know...

steven-murray commented:
Thanks @r-pascua for an amazing review. I've addressed all the easy stuff, so feel free to go through and resolve those things if they seem adequate. The stuff that's left is mostly the beam_id ordering, which I think we should chat live about.

r-pascua commented:
OK, I've resolved all the comments that I think have been addressed. By my account, it looks like these are the remaining issues:
* Update the `lsts` attribute to use a more robust method of retrieving the unique LSTs (the current implementation only works for one particular blt-order).
* Do some work to ensure that the simulated visibilities from `healvis` and `vis_cpu` are reshaped so that they match the shape of the underlying `data_array`. I would like to write a unit test for this, but I'm not sure about the best way to implement this. One thought that comes to mind is checking that redundant baselines give identical visibilities (up to some tolerance) when identical beams are used, and that all of the autocorrelations match.
* Make sure that the `write` method notes the version of `hera_sim` that was used (and ideally also the version(s) of the simulator(s) used).
* Figure out how to ensure that beams, antenna numbers, and so-on are all sorted in the same way. We should do this in a live collaborative session.

Thanks again for the great work! My hope is that we resolve most (if not all) of these issues during the live discussion. Once they're fixed, I'll give the green light for merging this.

r-pascua commented:
OK, if my accounting is correct, then we're very close to merging this. Here's what I think the remaining TODOs are:
* Update the `write` method to include the `hera_sim` version. (I feel like I'm missing something basic since I can't figure out how to do this in a non-circular way.)
* Update the beam handling to allow for `beam_ids` to be provided as an `array_like` object, subject to the constraint that it is either length 1 or has the same length as the number of active antennas (in which case we assume the IDs are sorted properly, and maybe issue a warning). This should fix the failing test.
* We still haven't gotten a review from @hughbg on whether the point source correction and polarization handling is correct.

Following this we have a few action items to complete before merging the overhaul into `main`. Here's my account of what we still need to complete after this PR is merged:
* Merge #170 
* Ensure that the new polarization features from #172 work with the refactor.
* Flesh out the unit tests to make sure we test that things work, and try to identify corner cases.

steven-murray commented:
OK @r-pascua:

1. The visibility simulator already writes the version of hera_sim to the history. I've added code now to write a nominal version of the simulator itself.
2. I've added logic for beam_ids
3. I think we can deal with polarization handling when we merge up the latest `main`. 

We should actually merge this now -- it's not going to `main` yet.
